samples/memcached-api-server/test.go

```diff
@@ -0,0 +1,50 @@
+package main
+
+import (
+	"log"
+	"runtime"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+)
+
+func main() {
+	_, filename, _, _ := runtime.Caller(0)
+	projectDir := filepath.Dir(filename)
+    runProjectTest(projectDir)
+}
+
+func runProjectTest(dir string) {
+	os.Setenv("TEST_ASSET_KUBECTL", "/tmp/kubebuilder/bin/kubectl")
+	os.Setenv("TEST_ASSET_KUBE_APISERVER", "/tmp/kubebuilder/bin/kube-apiserver")
+	os.Setenv("TEST_ASSET_ETCD", "/tmp/kubebuilder/bin/etcd")
+
+	log.Printf("Testing the project %s", dir)
+	cmds := []*exec.Cmd{
+		exec.Command("dep", "ensure"),
```

Comment:
WDYT of pulling these into libraries to make testing simpler with multiple projects.  I think we would be better served having multiple small to mid sized test projects instead of one sample project that tries to capture every possible usage.

Maybe create package: `internal/test/e2e/pkg` or something with:

```go
type KubebuilderTest struct {}

func  (KubebuilderTest) Generate() error {}
func  (KubebuilderTest) Docs(DocsOptions) error {}
func  (KubebuilderTest) Build(BuildOptions) error {}
func  (KubebuilderTest) Test(TestOptions) error {}
func  (KubebuilderTest) CreateConfig(ConfigOptions) error {}
```

Reply:
That is a good approach. I was thinking about similar stuff: putting the common used parts into the main testing files instead of in each project.

Reply:
Yes, we will be better served by having multiple of small projects, so optimizing for that usecase will help here.

@Liujingfang1 Not sure how much of the common stuff we will be able to put in the main test file because we have to support negative test cases as well. For ex. for RBAC validation test, I want to have a test project with missing RBAC annotations and `kubebuilder generate` will fail for that and that is expected.